megatron/core/datasets/data_schedule_utils.py

```
@@ -0,0 +1,885 @@
+            [[] for _ in range(total_gpus)],
+        )
+
+    def cp_min_fn(seq_len: int) -> int:
+        return dcp_gpus_needed(seq_len, max_seq_len_per_rank, min_cp_size)
+
+    def workload(seq_len: int, cp_size: int) -> float:
+        return (seq_len * seq_len) / cp_size
+
+    sample_seqlens = sorted(sample_seqlens, key=lambda x: x[1], reverse=True)
+    local_tall = sample_seqlens[0][1]
+    cap = float(local_tall) * float(max_seq_len_per_rank) * (1.0 + _DYNAMIC_CP_WORKLOAD_CAP_DELTA)
```

Ruff (0.15.21)

[error] Undefined name `_DYNAMIC_CP_WORKLOAD_CAP_DELTA`

(F821)